File: core/src/org/tetawex/ecf/model/RandomLevelFactory.java
package org.tetawex.ecf.model;

import com.badlogic.gdx.math.RandomXS128;
import com.badlogic.gdx.utils.OrderedSet;
import org.tetawex.ecf.util.IntVector2;
import org.tetawex.ecf.util.RandomProvider;

import java.util.function.*;

import java.util.*;

/**
 * Created by 1 on 06.06.2017.
 */
public class RandomLevelFactory
{
    private Cell[][] cellArray;
    private Set<Cell> cellSetAll;
    private Set<Cell> cellSetNoElements;
    private Set<Cell> cellSetWithElements;
    private RandomXS128 random;
    private int width,
                height;
    private int mana = 2;

    public RandomLevelFactory(int seed, int width, int height){
        RandomProvider.setSeed(seed);
        random = RandomProvider.getRandom();
        this.width = width;
        this.height = height;
        cellArray = CellArrayFactory.generateEmptyCellArray(width, height);
        createNullCells();
        createCellSets();
        fillTheBoard(width+height);
    }

    public int getMana(){
        return mana;
    }

    public Cell[][] getTheBoard(){
        return cellArray;
    }

    private void createNullCells(){

        int iterations = Math.min(width, height)+1;
        int x,y;
        for(int i =0; i<iterations;++i){
            if (random.nextBoolean()) {
                x = random.nextInt(width);
                y = random.nextInt(height);
                cellArray[x][y] = null;
            }
        }
    }

    private void createCellSets(){
        cellSetAll = new HashSet<Cell>();
        cellSetWithElements = new HashSet<Cell>();
        cellSetNoElements = new HashSet<Cell>();
        for (int x = 0; x < width; ++x){
            for (int y = 0; y < height; ++y){
                Cell cell = getCell(new IntVector2(x, y));
                if (cell != null) {
                    cellSetAll.add(cell);
                    cellSetNoElements.add(cell);
                }
            }
        }
    }

    private void fillTheBoard(int iterations){
        preFill();
        for(int i = 0; i < iterations; ++i){
            int rnd = random.nextInt(100);
            if (rnd < 20) {
                moveCell();
            }else if(rnd < 30){
                splitCell();
            }else if(rnd < 50){
                createOppositeElements();
            }else if(rnd < 100){
                createIdenticalElements();
            }
        }
    }

    private void preFill(){
        Cell cell1, cell2;
        cell1 = getRandomCellFromSet(cellSetAll);
        cell2 = getRandomNeighbour(cell1);
        cellSetNoElements.remove(cell1);
        cellSetWithElements.add(cell1);
        cellSetNoElements.remove(cell2);
        cellSetWithElements.add(cell2);
        cell1.getElements().add(Element.FIRE);
        cell2.getElements().add(Element.WATER);

        cell1 = getRandomCellFromSet(cellSetAll);
        cell2 = getRandomNeighbour(cell1);
        cellSetNoElements.remove(cell1);
        cellSetWithElements.add(cell1);
        cellSetNoElements.remove(cell2);
        cellSetWithElements.add(cell2);
        cell1.getElements().add(Element.LIGHT);
        cell2.getElements().add(Element.SHADOW);

        cell1 = getRandomCellFromSet(cellSetAll);
        cell2 = getRandomNeighbour(cell1);
        cellSetNoElements.remove(cell1);
        cellSetWithElements.add(cell1);
        cellSetNoElements.remove(cell2);
        cellSetWithElements.add(cell2);
        cell1.getElements().add(Element.EARTH);
        cell2.getElements().add(Element.AIR);
    }

    private void moveCell(){
        Cell from = getRandomCellFromSet(cellSetWithElements);
        if (from == null) return;
        List<Cell> neighbours = getNeighoubrsEmpty(from);
        if (neighbours.size() != 0){
            ++mana;
            Cell to = getRandom(neighbours);
            cellSetNoElements.remove(to);
            cellSetWithElements.add(to);
            cellSetWithElements.remove(from);
            cellSetNoElements.add(from);

            for(int i = 0; i < from.getElements().size; ++i){
                Element element = from.getElements().first();
                to.getElements().add(element);
                from.getElements().remove(element);
            }
        }
    }

    private void splitCell(){
        Cell from = null;
        for(int i = 0; i<5; ++i){
            from = getRandomCellFromSet(cellSetWithElements);
            if(from != null && from.getElements().size > 1) break;
        }
        if(from == null || from.getElements().size == 1) return;

        List<Cell> neighbours = getNeighoubrsEmpty(from);
        if (neighbours.size() != 0){
            ++mana;
            Cell to = getRandom(neighbours);
            cellSetNoElements.remove(to);
            cellSetWithElements.add(to);
            int amount = random.nextInt(from.getElements().size-1)+1;
            if (amount == from.getElements().size){
                cellSetWithElements.remove(from);
                cellSetNoElements.add(from);
            }
            for(int i = 0; i < amount; ++i){
                Element element = from.getElements().first();
                to.getElements().add(element);
                from.getElements().remove(element);
            }
        }
    }

    private void createOppositeElements(){
        Cell from = getRandomCellFromSet(cellSetNoElements);
        if (from == null) return;
        List<Cell> neighbours = getNeighoubrsEmpty(from);
        if (neighbours.size() != 0){

            Cell to = getRandom(neighbours);
            cellSetNoElements.remove(from);
            cellSetWithElements.add(from);
            cellSetNoElements.remove(to);
            cellSetWithElements.add(to);

            from.setElements(ElementFunctions.generateRandomElementSet());
            int size = from.getElements().size;
            if(mana > 1+size){
                mana-=size;
            }else{
                mana = 1;
            }
            for (Element element:
                 from.getElements()) {
                to.getElements().add(Element.getOpposite(element));
            }
        }
    }

    private void createIdenticalElements(){
        Cell from = getRandomCellFromSet(cellSetWithElements);
        if (from == null) return;
        List<Cell> neighbours = getNeighoubrsEmpty(from);
        if (neighbours.size() != 0){
            ++mana;
            Cell to = getRandom(neighbours);
            cellSetNoElements.remove(to);
            cellSetWithElements.add(to);
            for (Element element:
                    from.getElements()) {
                to.getElements().add(element);
            }
        }
    }

    private Cell getCell(IntVector2 pos){
        if (!(0 <= pos.x && pos.x < width)){
            return null;
        }
        if (!(0 <= pos.y && pos.y < height)){
            return null;
        }
        return cellArray[pos.x][pos.y];
    }

    public List<Cell> getNeighbours(Cell cell){
        List<Cell> res = new ArrayList<Cell>();
        IntVector2 pos = cell.getPosition();
        Cell neighbour;
        int x = pos.x;
        int y = pos.y;
        if(x%2 != 0){
            neighbour = getCell(new IntVector2(x-1, y+1));
            if (neighbour != null) res.add(neighbour);

            neighbour = getCell(new IntVector2(x+1, y+1));
            if (neighbour != null) res.add(neighbour);
        }else{
            neighbour = getCell(new IntVector2(x-1, y-1));
            if (neighbour != null) res.add(neighbour);

            neighbour = getCell(new IntVector2(x+1, y-1));
            if (neighbour != null) res.add(neighbour);
        }
        neighbour = getCell(new IntVector2(x-1, y));
        if (neighbour != null) res.add(neighbour);

        neighbour = getCell(new IntVector2(x+1, y));
        if (neighbour != null) res.add(neighbour);

        neighbour = getCell(new IntVector2(x, y+1));
        if (neighbour != null) res.add(neighbour);

        neighbour = getCell(new IntVector2(x, y-1));
        if (neighbour != null) res.add(neighbour);

        return res;
    }

    private List<Cell> getNeighoubrsEmpty(Cell cell){
        List<Cell> neighbours = getNeighbours(cell);
        List<Cell> res = new ArrayList<Cell>();
        for (Cell neighbour:
             neighbours) {
            if(neighbour.getElements().size == 0){
                res.add(neighbour);
            }
        }

        return res;
    }

    private List<Cell> getNeighboursNotEmpty(Cell cell){
        List<Cell> neighbours = getNeighbours(cell);
        List<Cell> res = new ArrayList<Cell>();
        for (Cell neighbour:
                neighbours) {
            if(neighbour.getElements().size != 0){
                res.add(neighbour);
            }
        }

        return res;
    }

    private Cell getRandom(List<Cell> cells){
        return cells.get(random.nextInt(cells.size()));
    }

    private Cell getRandomNeighbour(Cell cell){
        List<Cell> neighbours = getNeighbours(cell);
        return neighbours.get(random.nextInt(neighbours.size()));
    }

    private Cell getRandomCellFromSet(Set<Cell> set){
        if (set.size() == 0) return null;
        int index = random.nextInt(set.size());
        Iterator<Cell> iter = set.iterator();
        for (int i = 0; i < index; i++) {
            iter.next();
        }
        return iter.next();
    }
}

File: core/src/org/tetawex/ecf/model/RandomLevelFactory.kt
package org.tetawex.ecf.model

import com.badlogic.gdx.math.RandomXS128
import org.tetawex.ecf.util.IntVector2
import org.tetawex.ecf.util.RandomProvider

import java.util.*

/**
 * Created by 1 on 06.06.2017.
 */
class RandomLevelFactory(seed: Int, private val width: Int, private val height: Int) {
    val theBoard: Array<Array<Cell?>>
    private var cellSetAll: MutableSet<Cell?>? = null
    private var cellSetNoElements: MutableSet<Cell?>? = null
    private var cellSetWithElements: MutableSet<Cell?>? = null
    private val random: RandomXS128
    var mana = 2
        private set

    init {
        RandomProvider.setSeed(seed.toLong())
        random = RandomProvider.getRandom()
        theBoard = CellArrayFactory.generateEmptyCellArray(width, height)
        createNullCells()
        createCellSets()
        fillTheBoard(width + height)
    }

    private fun createNullCells() {

        val iterations = Math.min(width, height) + 1
        var x: Int
        var y: Int
        for (i in 0 until iterations) {
            if (random.nextBoolean()) {
                x = random.nextInt(width)
                y = random.nextInt(height)
                theBoard[x][y] = null
            }
        }
    }

    private fun createCellSets() {
        cellSetAll = HashSet()
        cellSetWithElements = HashSet()
        cellSetNoElements = HashSet()
        for (x in 0 until width) {
            for (y in 0 until height) {
                val cell = getCell(IntVector2(x, y))
                if (cell != null) {
                    cellSetAll!!.add(cell)
                    cellSetNoElements!!.add(cell)
                }
            }
        }
    }

    private fun fillTheBoard(iterations: Int) {
        preFill()
        for (i in 0 until iterations) {
            val rnd = random.nextInt(100)
            if (rnd < 20) {
                moveCell()
            } else if (rnd < 30) {
                splitCell()
            } else if (rnd < 50) {
                createOppositeElements()
            } else if (rnd < 100) {
                createIdenticalElements()
            }
        }
    }

    private fun preFill() {
        var cell1: Cell?
        var cell2: Cell?
        cell1 = getRandomCellFromSet(cellSetAll!!)
        cell2 = getRandomNeighbour(cell1)
        cellSetNoElements!!.remove(cell1)
        cellSetWithElements!!.add(cell1)
        cellSetNoElements!!.remove(cell2)
        cellSetWithElements!!.add(cell2)
        cell1!!.elements!!.add(Element.FIRE)
        cell2!!.elements!!.add(Element.WATER)

        cell1 = getRandomCellFromSet(cellSetAll!!)
        cell2 = getRandomNeighbour(cell1)
        cellSetNoElements!!.remove(cell1)
        cellSetWithElements!!.add(cell1)
        cellSetNoElements!!.remove(cell2)
        cellSetWithElements!!.add(cell2)
        cell1!!.elements!!.add(Element.LIGHT)
        cell2!!.elements!!.add(Element.SHADOW)

        cell1 = getRandomCellFromSet(cellSetAll!!)
        cell2 = getRandomNeighbour(cell1)
        cellSetNoElements!!.remove(cell1)
        cellSetWithElements!!.add(cell1)
        cellSetNoElements!!.remove(cell2)
        cellSetWithElements!!.add(cell2)
        cell1!!.elements!!.add(Element.EARTH)
        cell2!!.elements!!.add(Element.AIR)
    }

    private fun moveCell() {
        val from = getRandomCellFromSet(cellSetWithElements!!) ?: return
        val neighbours = getNeighoubrsEmpty(from)
        if (neighbours.size != 0) {
            ++mana
            val to = getRandom(neighbours)
            cellSetNoElements!!.remove(to)
            cellSetWithElements!!.add(to)
            cellSetWithElements!!.remove(from)
            cellSetNoElements!!.add(from)

            for (i in 0 until from.elements!!.size) {
                val element = from.elements!!.first()
                to.elements!!.add(element)
                from.elements!!.remove(element)
            }
        }
    }

    private fun splitCell() {
        var from: Cell? = null
        for (i in 0..4) {
            from = getRandomCellFromSet(cellSetWithElements!!)
            if (from != null && from.elements!!.size > 1) break
        }
        if (from == null || from.elements!!.size == 1) return

        val neighbours = getNeighoubrsEmpty(from)
        if (neighbours.size != 0) {
            ++mana
            val to = getRandom(neighbours)
            cellSetNoElements!!.remove(to)
            cellSetWithElements!!.add(to)
            val amount = random.nextInt(from.elements!!.size - 1) + 1
            if (amount == from.elements!!.size) {
                cellSetWithElements!!.remove(from)
                cellSetNoElements!!.add(from)
            }
            for (i in 0 until amount) {
                val element = from.elements!!.first()
                to.elements!!.add(element)
                from.elements!!.remove(element)
            }
        }
    }

    private fun createOppositeElements() {
        val from = getRandomCellFromSet(cellSetNoElements!!) ?: return
        val neighbours = getNeighoubrsEmpty(from)
        if (neighbours.size != 0) {

            val to = getRandom(neighbours)
            cellSetNoElements!!.remove(from)
            cellSetWithElements!!.add(from)
            cellSetNoElements!!.remove(to)
            cellSetWithElements!!.add(to)

            from.elements = ElementFunctions.generateRandomElementSet()
            val size = from.elements!!.size
            if (mana > 1 + size) {
                mana -= size
            } else {
                mana = 1
            }
            for (element in from.elements!!) {
                to.elements!!.add(Element.getOpposite(element))
            }
        }
    }

    private fun createIdenticalElements() {
        val from = getRandomCellFromSet(cellSetWithElements!!) ?: return
        val neighbours = getNeighoubrsEmpty(from)
        if (neighbours.size != 0) {
            ++mana
            val to = getRandom(neighbours)
            cellSetNoElements!!.remove(to)
            cellSetWithElements!!.add(to)
            for (element in from.elements!!) {
                to.elements!!.add(element)
            }
        }
    }

    private fun getCell(pos: IntVector2): Cell? {
        if (!(0 <= pos.x && pos.x < width)) {
            return null
        }
        return if (!(0 <= pos.y && pos.y < height)) {
            null
        } else theBoard[pos.x][pos.y]
    }

    fun getNeighbours(cell: Cell?): List<Cell?> {
        val res = ArrayList<Cell?>()
        val pos = cell!!.position
        var neighbour: Cell?
        val x = pos!!.x
        val y = pos.y
        if (x % 2 != 0) {
            neighbour = getCell(IntVector2(x - 1, y + 1))
            if (neighbour != null) res.add(neighbour)

            neighbour = getCell(IntVector2(x + 1, y + 1))
            if (neighbour != null) res.add(neighbour)
        } else {
            neighbour = getCell(IntVector2(x - 1, y - 1))
            if (neighbour != null) res.add(neighbour)

            neighbour = getCell(IntVector2(x + 1, y - 1))
            if (neighbour != null) res.add(neighbour)
        }
        neighbour = getCell(IntVector2(x - 1, y))
        if (neighbour != null) res.add(neighbour)

        neighbour = getCell(IntVector2(x + 1, y))
        if (neighbour != null) res.add(neighbour)

        neighbour = getCell(IntVector2(x, y + 1))
        if (neighbour != null) res.add(neighbour)

        neighbour = getCell(IntVector2(x, y - 1))
        if (neighbour != null) res.add(neighbour)

        return res
    }

    private fun getNeighoubrsEmpty(cell: Cell): List<Cell?> {
        val neighbours = getNeighbours(cell)
        val res = ArrayList<Cell?>()
        for (neighbour in neighbours) {
            if (neighbour!!.elements!!.size == 0) {
                res.add(neighbour)
            }
        }

        return res
    }

    private fun getNeighboursNotEmpty(cell: Cell): List<Cell?> {
        val neighbours = getNeighbours(cell)
        val res = ArrayList<Cell?>()
        for (neighbour in neighbours) {
            if (neighbour!!.elements!!.size != 0) {
                res.add(neighbour)
            }
        }

        return res
    }

    private fun getRandom(cells: List<Cell?>): Cell {
        return cells[random.nextInt(cells.size)]!!
    }

    private fun getRandomNeighbour(cell: Cell?): Cell? {
        val neighbours = getNeighbours(cell)
        return neighbours[random.nextInt(neighbours.size)]
    }

    private fun getRandomCellFromSet(set: Set<Cell?>): Cell? {
        if (set.size == 0) return null
        val index = random.nextInt(set.size)
        val iter = set.iterator()
        for (i in 0 until index) {
            iter.next()
        }
        return iter.next()
    }
}
